File: code_python/utils/generate_figures.py
"""
Author: Rayla Kurosaki

GitHub: https://github.com/rkp1503
"""

import os

import matplotlib.pyplot as plt


def save_figure(filename: str):
    path_to_project_dir: str = os.path.dirname(os.getcwd())
    path_to_report_dir: str = os.path.join(path_to_project_dir, "report")
    path_to_figs_dir: str = os.path.join(path_to_report_dir, "figures")
    if os.path.exists(path_to_figs_dir):
        path_to_png_dir: str = os.path.join(path_to_figs_dir, "png")
        if not os.path.exists(path_to_png_dir):
            os.makedirs(path_to_png_dir)
            pass
        path_to_png: str = os.path.join(path_to_png_dir, filename)
        plt.savefig(f"{path_to_png}.png", bbox_inches="tight")
        path_to_eps_dir: str = os.path.join(path_to_figs_dir, "eps")
        if not os.path.exists(path_to_eps_dir):
            os.makedirs(path_to_eps_dir)
            pass
        path_to_eps: str = os.path.join(path_to_eps_dir, filename)
        plt.savefig(f"{path_to_eps}.eps", bbox_inches="tight")
        pass
    return None


def my_plot(ts, solutions, title, xaxis, yaxis, legend_data, norm=1,
            detailed_fig=False, filename=""):
    graphs = []
    for (i, style) in enumerate(legend_data.values()):
        label, color = style
        temp_graph = plt.plot(ts, solutions[:, i] / norm, label=label,
                              color=color, linewidth=1)
        graphs += temp_graph
        pass
    labels = [graph.get_label() for graph in graphs]
    plt.title(title)
    plt.xlabel(xaxis)
    plt.ylabel(yaxis)
    if detailed_fig:
        plt.legend(graphs, labels, bbox_to_anchor=(1.01, 1.015),
                   loc='upper left')
        plt.minorticks_on()
        plt.grid(which='major', linestyle='-', linewidth='0.5',
                 color='#000000')
        plt.grid(which='minor', linestyle=':', linewidth='0.5',
                 color='#FF0000')
        pass
    else:
        plt.legend(graphs, labels)
        pass
    if filename != "":
        save_figure(filename)
        pass
    plt.show()
    return None


def my_3D_phase_portrait(sol, title, xaxis, yaxis, zaxis, detailed_fig=False,
                         filename=""):
    ax = plt.axes(projection='3d')
    ax.plot3D(sol[:, 0], sol[:, 1], sol[:, 2])
    plt.title(title)
    ax.set_xlabel(xaxis)
    ax.set_ylabel(yaxis)
    ax.set_zlabel(zaxis)
    if detailed_fig:
        plt.minorticks_on()
        plt.grid(which='major', linestyle='-', linewidth='0.5',
                 color='#000000')
        plt.grid(which='minor', linestyle=':', linewidth='0.5',
                 color='#FF0000')
        pass
    if filename != "":
        save_figure(filename)
        pass
    plt.show()
    return None


def my_2D_phase_portrait(solution_x, solution_y, title, xaxis, yaxis,
                         detailed_fig=False, filename=""):
    plt.plot(solution_x, solution_y, linewidth=1)
    plt.title(title)
    plt.xlabel(xaxis)
    plt.ylabel(yaxis)
    if detailed_fig:
        plt.minorticks_on()
        plt.grid(which='major', linestyle='-', linewidth='0.5',
                 color='#000000')
        plt.grid(which='minor', linestyle=':', linewidth='0.5',
                 color='#FF0000')
        pass
    if filename is not None:
        save_figure(filename)
        pass
    plt.show()
    return None


def my_bifurcation_diagram(data_stable, data_unstable, variable, parameter,
                           color, detailed_fig=False, filename=""):
    if data_stable == []:
        parameters_stable, solution_stable = [], []
        pass
    else:
        parameters_stable, solution_stable = data_stable[0], data_stable[1]
        pass
    if data_stable == []:
        parameters_unstable, solution_unstable = [], []
        pass
    else:
        parameters_unstable, solution_unstable = data_unstable[0], data_unstable[1]
        pass
    plt.title(f"Density of Species ${variable}$ vs ${parameter}$")
    plt.xlabel(f"${parameter}$")
    plt.ylabel(f"Density of Species ${variable}$")
    plt.scatter(parameters_stable, solution_stable, s=1, color=color,
                label="Stable")
    plt.scatter(parameters_unstable, solution_unstable, s=1, color="g",
                label="Unstable")
    plt.legend()
    if filename is not None:
        save_figure(filename)
        pass
    plt.show()
    return None

File: code_julia/utils/GenerateFigures.jl
#=============================================================================
Author: Rayla Kurosaki
GitHub: https://github.com/rkp1503
=============================================================================#

module GenerateFigures

import Plots: @layout, plot, plot!, scatter, scatter!

function my_plot(solutions, title, xaxis, yaxis, legend_data; norm=1, detailed_fig=false, export_fig=false)
    figure = plot(title=title, xaxis=xaxis, yaxis=yaxis, gridalpha=0)
    if detailed_fig
        plot!(legend=:outertopright, background_color_inside="#DDDDDD", gridalpha=1, gridlinewidth=:1, foreground_color_grid="#000000", minorgrid=true, minorgridalpha=0.5, foreground_color_minor_grid=:"#FF0000")
    end
    for (i, style) in enumerate(collect(values(legend_data)))
        label, color = style
        plot!(figure, solutions.t, solutions'[:, i]/norm, label=label, linecolor=color, linewidth=1.5)
    end
    if export_fig
    end
    return figure
end;

function my_3D_phase_portrait(solutions, title, xaxis, yaxis, zaxis; detailed_fig=false, export_fig=false)
    figure = plot(title=title, xaxis=xaxis, yaxis=yaxis, zaxis=zaxis, legend=false, gridalpha=0)
    if detailed_fig
        plot!(background_color_inside="#DDDDDD", gridalpha=1, gridlinewidth=:1, foreground_color_grid="#000000", minorgrid=true, minorgridalpha=0.5, foreground_color_minor_grid=:"#FF0000")
    end
    plot!(figure, solutions'[:, 1], solutions'[:, 2], solutions'[:, 3], linewidth=1.5)
    if export_fig
    end
    return figure
end;

function my_2D_phase_portrait(solution_x, solution_y, title, xaxis, yaxis; detailed_fig=false, export_fig=false)
    figure = plot(title=title, xaxis=xaxis, yaxis=yaxis, legend=false, gridalpha=0)
    if detailed_fig
        plot!(background_color_inside="#DDDDDD", gridalpha=1, gridlinewidth=:1, foreground_color_grid="#000000", minorgrid=true, minorgridalpha=0.5, foreground_color_minor_grid=:"#FF0000")
    end
    plot!(figure, solution_x, solution_y, linewidth=1.5)
    if export_fig
    end
    return figure
end;

function my_phase_portraits(solutions, xaxis, yaxis, zaxis)
    solution_x = solutions'[:, 1]
    solution_y = solutions'[:, 2]
    solution_z = solutions'[:, 3]
    fig_a = my_3D_phase_portrait(solutions, "3D Phase Portrait", xaxis, yaxis, zaxis; detailed_fig=true)
    fig_b = my_2D_phase_portrait(solution_x, solution_y, "Species X and Y", xaxis, yaxis; detailed_fig=true)
    fig_c = my_2D_phase_portrait(solution_x, solution_z, "Species X and Z", xaxis, zaxis; detailed_fig=true)
    fig_d = my_2D_phase_portrait(solution_y, solution_z, "Species Y and Z", yaxis, zaxis; detailed_fig=true)
    l = @layout [a b ; c d]
    return plot(fig_a, fig_b, fig_c, fig_d, layout=l)
end;

function my_bifurcation_diagram(data_stable, data_unstable, variable, parameter, color; detailed_fig=false, export_fig=false)
    if data_stable == []
        parameters_stable, solution_stable = [], []
    else
        parameters_stable, solution_stable = data_stable[:, 1], data_stable[:, 2]
    end
    if data_unstable == []
        parameters_unstable, solution_unstable = [], []
    else
        parameters_unstable, solution_unstable = data_unstable[:, 1], data_unstable[:, 2]
    end
    figure = scatter(title="Density of Species $(variable) vs $(parameter)", xaxis="$(parameter)", yaxis="Density of Species $(variable)")
    if detailed_fig
        scatter!(legend=:outertopright, background_color_inside="#DDDDDD", gridalpha=1, gridlinewidth=:1, foreground_color_grid="#000000", minorgrid=true, minorgridalpha=0.5, foreground_color_minor_grid=:"#FF0000")
    end
    scatter!(figure, parameters_stable, solution_stable, markersize=1, markercolor=color, markerstrokecolor=color, label="Stable")
    scatter!(figure, parameters_unstable, solution_unstable, markersize=1, markercolor=:lime, markerstrokecolor=:lime, label="Unstable")
    if export_fig
    end
    return figure
end;

function my_bifurcation_diagrams(solutions, data, variables, bifurcartion_parameter, legend_data)
    x_arr_t, x_arr_s, x_arr_u, y_arr_t, y_arr_s, y_arr_u, z_arr_t, z_arr_s, z_arr_u = data
    x, y, z = variables
    title = "Time Evolution of Each Species"
	xaxis = "Time in Days"
	yaxis = "Population size"
    fig_a = my_plot(solutions, title, xaxis, yaxis, legend_data; detailed_fig=true)
    fig_b = my_bifurcation_diagram(x_arr_s, x_arr_u, x, bifurcartion_parameter, "black"; detailed_fig=true)
    fig_c = my_bifurcation_diagram(y_arr_s, y_arr_u, y, bifurcartion_parameter, "red"; detailed_fig=true)
    fig_d = my_bifurcation_diagram(z_arr_s, z_arr_u, z, bifurcartion_parameter, "blue"; detailed_fig=true)
    l = @layout [a b ; c d]
    return plot(fig_a, fig_b, fig_c, fig_d, layout=l)
end;
    
end